### ISSUE TITLE
Add Cash Address Support (sending only).  

### ISSUE THREAD

**Issue by HashEngineering** [bitcoincash-wallet/bitcoincash-wallet#9]
This includes scan and paste support.

**Comment by HashEngineering** [comment]
This will also include adding to the Address Book.

@snapfractalpop - if you are interested, I have a test version of the app that supports the new address format for scanning and pasting into the app for Send and the Address Book.

**Comment by snapfractalpop** [comment]
Is it on a particular branch that I can checkout?

**Comment by HashEngineering** [comment]
There will be after I push it both for this repo and bitcoinj-cash.

**Comment by HashEngineering** [comment]
**bitcoincash-wallet:**
https://github.com/bitcoincash-wallet/bitcoincash-wallet/tree/cashaddr/wallet

**bitcoinj:**
https://github.com/bitcoincash-wallet/bitcoinj/tree/rseibane-cash

**Comment by snapfractalpop** [comment]
Thanks, I'll pull them and have a look.

**Comment by snapfractalpop** [comment]
Is the regex on [this line](https://github.com/bitcoincash-wallet/bitcoincash-wallet/compare/master...bitcoincash-wallet:cashaddr#diff-d793d9ec2102c9c94fa3a9d94acfb26bR390) just a temporary stub? I noticed it's missing the characters ```['b', 'i', 'o', '1']```.

**Comment by HashEngineering** [comment]
https://github.com/bitcoincashorg/spec/blob/master/cashaddr.md

Yes, that line is from the spec.

**Comment by snapfractalpop** [comment]
I'm having trouble adding the local bitcoinj-cash dependency. I have cloned it from git and have tried to import it as a module, and also tried with bitcoinj as a subdirectory within the bitcoincash-wallet project. Is mvn required in order to use the local version (0.15-SNAPSHOT)?

**Comment by HashEngineering** [comment]
I use maven.

`mvn clean install -DskipTests` will install it under your local maven repository. 

`mavenLocal();` is in the build.gradle of the wallet so gradle will look in the local maven repo.

**Comment by snapfractalpop** [comment]
I have compiled and installed the changes in this branch, and so far, everything is working as expected.

One thing I did notice is that when cashaddr addresses are used, they are converted to "legacy" addresses in the "pay to" field of the payment form. From a UX stand point, I wonder if this might confuse users (as the letters and numbers in the base32 and base58 encodings differ).

Maybe we can display both formats, with one as parenthetical. Perhaps the format shown as the "main" one could be based on a user setting after implementing full support for cashaddr. Any thoughts on this?

**Comment by HashEngineering** [comment]
Thanks for testing this app.  For this version, by design, the behavior is as you pointed out.  The rest of the app still expects to use Legacy format addresses, so I had the app convert cash address to Legacy.  The send page for instance will use that legacy address when building the transaction.

I am sure this will be confusing to some users, but at least it would work to pay to cash addresses.

I was hoping to display the address is the same format it was scanned in, and there might be an easy way to do that without breaking the send functionality.  Therefore, I will make some changes to remove the confusion.

A later update would include full cash address support in all the screens of the app.  I am not so sure about allowing users to choose the format that they want to see.  While Bitcoin-ABC has a command line setting, I have seen many apps and websites use the cash addr format.  Going forward all the addresses would be in cash addr format.

**Comment by HashEngineering** [comment]
@snapfractalpop I pushed an update to https://github.com/bitcoincash-wallet/bitcoincash-wallet/tree/cashaddr which will display the same format as scanned or pasted for the send screen.

**Comment by snapfractalpop** [comment]
I agree that going forward, all addresses should be in cashaddr format. I also tend to agree about not having a user setting. Some things that come to mind that I think can be addressed (no pun intended):

- when a user scans/pastes a legacy address, they expect to see the same letters and numbers (this is currently satisfied)
- when a user scans/pastes a cashaddr address, they expect to see the same letters and numbers (without concern for case)

The following relate more to #10:

- when a user wants to receive coins from a peer whose wallet only supports the cashaddr format, they expect that the cashaddr address can be shown
- when a user wants to receive coins from a peer whose wallet only supports the legacy format, they expect that the legacy address can be shown (this is currently satisfied)

Reflecting on this a bit, I think these can all be satisfied without requiring a user setting. In fact, I think a setting makes the UX cumbersome (and the program unnecessarily stateful). Maybe for sending coins, we can show the legacy address small and grayed out a bit, and for accepting coins, we can toggle the formats in the display of the QR (and text field) gracefully as is done [here](https://cashaddr.bitcoincash.org/).

Do you think these use cases can and should be supported? Any alternative ways to achieve the same? Thanks for being so responsive on these matters.

**Comment by HashEngineering** [comment]
The first two items you list are intended and hopefully present in my latest code updates.

The second two items, I don't think they are possible.  The receiving wallet does not know or cannot tell what address format the sending wallet is using.  Transactions don't actually contain addresses, but instead public key hashes (p2pkh).  That format is rendered as an human readable address by the software, but the p2pkh doesn't contain information on how to format.

> Do  you think these use cases can and should be supported? 
Yes and Yes for the first two.  No and Yes it would be cool for the second two.

> Any alternative ways to achieve the same? 
For the second two, one possibility is a setting on the request coins screen to change the address format.  But, I don't think the wallet file itself supports saving the format of the address to display.  Such information would be lost between screens or when the app is restarted.  The scan or send process can show the same format, but once the user goes back to the main screen, that information is lost.  Not sure if there is a way around this at this time.

> Thanks  for being so responsive on these matters.
Thank's for the feedback, it helps with making this a better app.

**Comment by snapfractalpop** [comment]
I'm aware that the underlying public key is no different, regardless of the format used to display it. To further clarify the use cases I'm describing in those second two bullets, I'm mostly referring to the scenario where two peers are meeting in person.

Alice uses our app, and Bob uses either "legacy only app" or "cashaddr only app". Currently, if Bob uses a "legacy only app", Alice can show Bob a QR code for payment, and Bob has no problem paying it (his app "speaks" legacy). So the last bullet point is satisfied.

I imagine when full support for the cashaddr format is rolled out, the 3rd bullet point would be satisfied. I.e. Alice can show a cashaddr format QR to Bob, and Bob can use his "cashaddr only app" to pay it.

So, aside from some of the validation / error detection that may be enabled by the cashaddr format, the format feels like something that belongs at the view level. In a sense, the different formats are really just different views of the same model. I think it would be the best UX if backward compatibility is maintained while we wait for all the Bob's out there to upgrade or install an app that supports cashaddr too. But really, this part of the discussion probably belongs on #10, as you've already handled the first two points.

I haven't yet tested the code, btw, as I wrote the last comment before realizing you had sent that commit. You are fast!

**Comment by snapfractalpop** [comment]
I just realized that you wrote more in the quotes (I skipped over those thinking they were only what I wrote before).

> For the second two, one possibility is a setting on the request coins screen to change the address format. But, I don't think the wallet file itself supports saving the format of the address to display. Such information would be lost between screens or when the app is restarted. The scan or send process can show the same format, but once the user goes back to the main screen, that information is lost. Not sure if there is a way around this at this time.

I think that it would be perfectly fine to not save the format used for a particular request screen. I simply want to enable the most versatility "in the wild" while interacting with other people with apps of various incompatibilities. So, my thinking is to basically have something as simple as the view toggle example at https://cashaddr.bitcoincash.org/ (btw, in the case of that page, you have to actually enter a valid address to see the example I mean).

**Comment by snapfractalpop** [comment]
I know you already implemented a solution for ensuring that the format type matches that which the user used as the receiving address in the send fragment, but [here is an example](https://gist.github.com/snapfractalpop/56c3bdf5140dd0f61b5d857a6c50a11a) of what I meant in https://github.com/bitcoincash-wallet/bitcoincash-wallet/issues/9#issuecomment-372530177.

This is by no means complete (I think the text overflows out of the view, for example). I'm mentioning it only to better convey the idea.

**Comment by HashEngineering** [comment]
Thanks for that code example.  I understand what you mean by it.

I think what I will do is:
1) perform more testing on current version we have looked at here.
2) if its working, then publish to google play and GitHub
3) begin thinking about #10 

Thank you for testing and your suggestions.

**Comment by snapfractalpop** [comment]
You're welcome. Thanks for the timely responses and active development!

**Comment by HashEngineering** [comment]
This has been released for a week or so and no complaints about bugs.

**Comment by HashEngineering** [comment]
2 user reports that new BCH addresses cannot be read, but there is not enough information to determine the cause of the failure.

**Comment by snapfractalpop** [comment]
Hm.. I have not had any problems with it, but then again I haven't tested all the edge cases. Did they report it on the Play Store? Is there any way you can ask them about the failure (error message, etc.)?

**Comment by HashEngineering** [comment]
It was reported through the app using email.  There are two users, but there is much difficulty understanding what they are talking about.  I will ask for screenshots as that will be worth 1000 words.

**Comment by snapfractalpop** [comment]
That's a good idea.

**Comment by imaginaryusername** [comment]
It seems like the wallet's receiving format (a weird combination of new prefix and legacy address, bitcoincash:1xxxxx) was not recognized by itself. If I copy from the wallet and paste straight into send, it says "invalid". Please fix.
